akka-stream/src/main/scala/akka/stream/impl/io/ByteStringParser.scala

```diff
@@ -72,8 +89,21 @@ private[akka] abstract class ByteStringParser[T] extends GraphStage[FlowShape[By
           if (acceptUpstreamFinish) completeStage()
           else current.onTruncation()
         } else pull(bytesIn)
+
+        DontRecurse
+      }
+
+    @tailrec private def doParse(remainingRecursions: Int = recursionLimit): Unit =
```

Comment:
What do we gain by splitting up the function into these two parts? For me it looks cleaner when there is no additional Boolean but the recursion is inlined.